<center><img src="https://github.com/insaid2018/Term-1/blob/master/Images/INSAID_Full%20Logo.png?raw=true" width="240" height="100" /></center>

**<center><h3>Data Manipulation with Pandas Assignment Solution</h3></center>**

In [1]:
import numpy as np
import pandas as pd

---
**<h4>Question 1:** Create a function that converts the each letter of first word in a series to uppercase.</h4>

- For example, you passed the series in left column and return the values as shown in right column:

|Original Series|Transformed Series
|:--|:--|
|how|How|
|to|To|
|learn|Learn|
|data science?|Data science?|

<details>

**<summary>Hint:</summary>**

- You can achieve this question in many ways.

- One way is to perform mapping over the series and using the capitalize functionality of python.

</details>

In [2]:
def upperSeries(series=None):
  """Converts the each letter of first word in a series to uppercase."""

  # Performing mapping over the series elements
  series = series.map(lambda x: x.capitalize())
  
  return series

In [3]:
upperSeries(series=pd.Series(['how', 'to', 'learn', 'data science?']))

0              How
1               To
2            Learn
3    Data science?
dtype: object

---
**<h4>Question 2:** Create a function that takes dictionary as input and returns the rows of dataframe containing null data.</h4>

- For example, you passed the dictionary and dataframe looks like as shown below:

|#|Animal|Age|Visits|Priority|
|:--|:--|:--|:--|:--|
|01|dog|NaN|3|yes|
|02|dog|5|2|no|
|03|cat|2|3|no|
|04|snake|4|1|no|
|05|cat|NaN|1|yes|

- The output data will be as shown below:

|#|Animal|Age|Visits|Priority|
|:--|:--|:--|:--|:--|
|01|dog|NaN|3|yes|
|05|cat|NaN|1|yes|

<details>

**<summary>Hint:</summary>**

- Pass the dictionary containing data to the function.

- Create a dataframe using the pandas DataFrame functionality.

- Filter the data rows containing null data using isnull() and any() method.

- You can also try out some condition inside the dataframe index brackets.

- Return the filtered data back.

</details>

In [4]:
def showNanRows(data_dict=None):
  """Returns the data rows containing null data."""

  # Create a dataframe out of the data dictionary passed
  data_frame = pd.DataFrame(data=data_dict)

  # Filter the data rows containing null data
  data_frame = data_frame[data_frame.isnull().any(axis=1)]

  return data_frame

In [5]:
data_dict = {'Animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
             'Age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3], 
             'Visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
             'Priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

showNanRows(data_dict=data_dict)

Animal  Age  Visits Priority
3    dog  NaN       3      yes
7    cat  NaN       1      yes

---
**<h4>Question 3:** Create a function that takes dictionary as input and returns the dataframe of grouped mean of a certain column.</h4>

- For example, you passed the dictionary and dataframe looks like as shown below:

|#|Animal|Age|Visits|Priority|
|:--|:--|:--|:--|:--|
|01|dog|NaN|3|yes|
|02|dog|5|2|no|
|03|cat|2|3|no|
|04|snake|4|1|no|
|05|cat|NaN|1|yes|

- The output data that has been grouped by Animal of mean Age will be as shown below:

|#|Animal|Age|
|:--|:--|:--|
|01|cat|2|
|02|dog|5|
|03|snake|4|

<details>

**<summary>Hint:</summary>**

- Pass the dictionary containing data to the function.

- Create a dataframe using the pandas DataFrame functionality.

- Groupby data using pandas package according to the specified groupbycol value.

- Extract columns only specified in returncollist parameter.

- Return the output data.

</details>

In [6]:
def showGroupByMean(data_dict=None, groupbycol=None, returncollist=None):
  """Returns the dataframe of grouped mean of animal column."""

  # Create a dataframe out of the data dictionary passed
  data_frame = pd.DataFrame(data=data_dict)

  # Groupby data with the specified groupby column and columns to return in output
  grouped_mean = data_frame.groupby(groupbycol)[returncollist].mean()

  return grouped_mean

In [7]:
data_dict = {'Animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
             'Age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3], 
             'Visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
             'Priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

showGroupByMean(data_dict=data_dict, groupbycol='Animal', returncollist=['Age'])

Age
Animal     
cat     2.5
dog     5.0
snake   2.5

---
**<h4>Question 4:** Create a function that returns the indexes of the dataframe if values matched at a certain index.</h4>

- **Note:** You can create two columns with random values or labels within a certain range.

- For example, you initialized dataframe columns with random values as shown below:

|#|Fruit 1|Fruit 2|
|:--|:--|:--|
|01|Banana|Orange|
|02|Apple|Orange|
|03|Apple|Apple|
|04|Orange|Orange|
|05|Banana|Apple|

- The returned output will be `[3, 4]`.

<details>

**<summary>Hint:</summary>**

- Pass a list of items that will be used to create a dataframe of specified data rows.

- To achieve this objective you can use numpy's random choice functionality.

- Next, you can use numpy's where() method to extract out the indexes of data rows containing same values.

- Return the extracted indexes.

</details>

In [8]:
def checkIndex(itemlist=None, datasize=None):
  """Returns indexes of data rows where both values are same."""

  # Create a dataframe with two columns having value in certain range of list.
  df = pd.DataFrame(data = {'Column 1': np.random.choice(a=itemlist, size=datasize),
                            'Column 2': np.random.choice(a=itemlist, size=datasize)})
  
  # Extract out the indexes where data values match
  index = np.where(df['Column 1'] == df['Column 2'])

  # Display the output
  return index

In [9]:
checkIndex(itemlist=['apple', 'orange', 'banana'], datasize=10)

(array([0, 3, 9]),)

---
**<h4>Question 5:** Create a function that returns the indexes of the series if values present are multiple of a certain value.</h4>

- For example, you created a random series of integer as shown below:

|#|Series|
|:--:|:--:|
|01|8|
|02|9|
|03|8|
|04|1|
|05|8|

- The value to be checked as the multiple of values present in the series is `4`.

- The returned output will be `[1, 3, 5]`.

<details>

**<summary>Hint:</summary>**

- Pass a list of items that will be used to create a dataframe of specified data rows.

- To achieve this objective you can use numpy's random choice functionality.

- Next, you can use numpy's where() method to extract out the indexes of data rows containing same values.

- Return the extracted indexes.

</details>

In [10]:
def checkMultiple(multiple=None, randstart=None, randstop=None, totalrecords=None):
  """Returns the indexes of the series if values present are multiple of a certain value."""

  # Generate random integers according to the specifed parameters
  data = np.random.randint(low=randstart, high=randstop, size=totalrecords)

  # Create a pandas series of random integers with specified values in function
  series = pd.Series(data=data)

  # Extract the index position where series value is multiple of passed value
  pos = np.where(series%multiple == 0)

  return pos

In [11]:
checkMultiple(multiple=4, randstart=1, randstop=10, totalrecords=10)

(array([3, 4, 5, 6, 8]),)

---
**<h4>Question 6:** Create a function that takes two series and stacks them in either horizontal way or vertical way.</h4>

- For example you entered following two series:

|Series 1|
|:--:|
|0|
|1|
|2|
|3|
|4|
|5|

|Series 2|
|:--:|
|I|
|N|
|S|
|A|
|I|
|D|

- The returned output will be as shown below if the option selected is horizontal:

|Series 1|Series 2|
|:--:|:--:|
|0|I|
|1|N|
|2|S|
|3|A|
|4|I|
|5|D|


<details>

**<summary>Hint:</summary>**

- Use conditional opertaion to check the passed option of stacking the list.

- If the option is horizontal, then perform concatenation of series along axis 1.

- If the option is verticle, then perform appending of series.

- Return the final result.

</details>

In [12]:
def stackSeries(series1=None, series2=None, stack='verticle'):
  """Return series stacked in horizontal way."""

  # Check if option selected is horizontal or not
  if stack == 'horizontal':
    return pd.concat([series1, series2], axis=1)

  return series1.append(series2)

In [13]:
s1 = pd.Series(range(6))
s2 = pd.Series(list('INSAID'))

stackSeries(series1=s1, series2=s2, stack='horizontal')

0  1
0  0  I
1  1  N
2  2  S
3  3  A
4  4  I
5  5  D

---
**<h4>Question 7:** Create a function that performs min-max scaling over the numerical columns of the data.</h4>

- **Dataset link:** 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

- For example, you have a column where values are `[4.9, 4.7, 4.6, 5.0, 5.4]`.

- After performing min-max scaling the values will become `[0.22, 0.16, 0.11, 0.08, 0.19]`.

<details>

**<summary>Hint:</summary>**

- Read the data from the specified link and provide columns names if not present already.

- You can use pandas read_csv functionality and its constituent parameters.

- Extract and append numerical column names out of the dataset using control flow & string operations.

- Again start a new control flow operation and extract minimum and maximum value of each column.

- Use apply() method and lambda operation over each column to perform min-max scaling.

- The formula of min-max scaling is: $\large\frac {value - min}{max - min}$

- One way is to perform mapping over the series and using the capitalize functionality of python.

</details>

In [14]:
def minMaxScaling(link=None, datacolnames=None):
  """Performs min-max scaling over the numerical features."""

  # Read the data from the specified link and column names if not present already
  data = pd.read_csv(filepath_or_buffer=link, names=datacolnames)

  # Extract the numerical features
  numfeatures = []
  for i in data.columns:
    if (data[i].dtype == float) | (data[i].dtype == int):
      numfeatures.append(i)

  # Iterate scaling over each numerical column
  for i in numfeatures:

    # Get minimum value of the column
    feat_min = data[i].min()

    # Get maximum value of the column
    feat_max = data[i].max()

    # Perform min-max scaling over each numerical column
    data[i] = data[i].apply(lambda x: (x - feat_min) / (feat_max - feat_min))
  
  return data

In [15]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
names=['sepal length', 'sepal width', 'petal length', 'petal width', 'class']

minMaxScaling(link=url, datacolnames=names)

sepal length  sepal width  petal length  petal width           class
0        0.222222     0.625000      0.067797     0.041667     Iris-setosa
1        0.166667     0.416667      0.067797     0.041667     Iris-setosa
2        0.111111     0.500000      0.050847     0.041667     Iris-setosa
3        0.083333     0.458333      0.084746     0.041667     Iris-setosa
4        0.194444     0.666667      0.067797     0.041667     Iris-setosa
..            ...          ...           ...          ...             ...
145      0.666667     0.416667      0.711864     0.916667  Iris-virginica
146      0.555556     0.208333      0.677966     0.750000  Iris-virginica
147      0.611111     0.416667      0.711864     0.791667  Iris-virginica
148      0.527778     0.583333      0.745763     0.916667  Iris-virginica
149      0.444444     0.416667      0.694915     0.708333  Iris-virginica

[150 rows x 5 columns]

---
**<h4>Question 8:** Create a function that returns the index positions of the missing data value and its total in a specific column.</h4>

- **Dataset link:** 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

- For example, you have a column where values are `[4.9, 4.7, 4.6, 5.0, 5.4]`.

- After performing min-max scaling the values will become `[0.22, 0.16, 0.11, 0.08, 0.19]`.
  
<details>

**<summary>Hint:</summary>**

- Read the data from the specified link and provide columns names if not present already.

- You can use pandas read_csv functionality and its constituent parameters.

- Calculate total missing values in a specified column using pandas isnull() and sum() method.

- Similarly, you can use numpy's where() method in addtion to the isnull() method to get the index positions of missing data.

</details>

In [16]:
def showMissingIndex(link=None, datacolnames=None, checkcol=None):
  """Returns index positions of missing data row of a specific column."""

  # Read the data from the specified link and column names if not present already
  data = pd.read_csv(filepath_or_buffer=link, names=datacolnames)

  # Display total missing values in a specified column
  print('Total Missing:', data[checkcol].isnull().sum())
  
  # Display the index positions of missing data row
  print('Indexes:', np.where(data[checkcol].isnull()))

In [17]:
url = 'https://raw.githubusercontent.com/insaid2018/Domain_Case_Studies/master/Healthcare/Heart%20Disease%20Diagnosis/Data/heart_disease.csv'

showMissingIndex(link=url, checkcol='cholesterol')

Total Missing: 18
Indexes: (array([ 38,  45,  46,  55, 125, 162, 164, 179, 199, 204, 221, 232, 237,
       249, 284, 303, 314, 328]),)


---
**<h4>Question 9:** Create a function that replace missing values with the mean value of the column, if present.</h4>

- For example, you have a following dataset:

|#|age|sex|chest_pain_type|resting_blood_pressure|cholesterol|fasting_blood_sugar|rest_ecg|target|
|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
|01|63|1|asymptomatic	|145|233.0|1|1|1|
|02|37|0|non-anginal pain|140|245.0|0|1|0|
|03|41|1|typical anginia	|135|NaN|1|1|0|
|04|48|0|asymptomatic	|110|274|0|0|0|
|05|66|1|non-anginal pain|130|288|1|1|1|

- The cholesterol column contains one missing value and mean of this column is `262.25`.

- The final dataframe will look something like as follows:

|#|age|sex|chest_pain_type|resting_blood_pressure|cholesterol|fasting_blood_sugar|rest_ecg|target|
|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
|01|63|1|asymptomatic	|145|233.0|1|1|1|
|02|37|0|non-anginal pain|140|245.0|0|1|0|
|03|41|1|typical anginia	|135|262.25|1|1|0|
|04|48|0|asymptomatic	|110|274|0|0|0|
|05|66|1|non-anginal pain|130|288|1|1|1|

<details>

**<summary>Hint:</summary>**

- Read the data from the specified link and provide column names if not present already.

- You can use pandas read_csv functionality and its constituent parameters.

- Check if the data column contains missing data using pandas isnull() and any() method.

- If column contains missing data then replace the missing data cell using .replace() method.

- After replacing the missing data values in the column again check if there exists missing data.

- If no missing data in the column then return the final data.

</details>

In [18]:
def replaceMissingMean(link=url, datacolnames=None, colname=None):
  """Imputes missing data value with mean value of the column."""

  # Read the data from the specified link and column names if not present already
  data = pd.read_csv(filepath_or_buffer=link, names=datacolnames)

  # Display if originally column contains missing values 
  print('Before Replacing, Contains Missing?:', data[colname].isnull().any())

  # Replace missing data value
  data[colname] = data[colname].replace(to_replace=np.nan, value=data[colname].mean())

  # Display if after replacing column contains missing values 
  print('After Replacing, Contains Missing?:', data[colname].isnull().any())

  return data

In [19]:
url = 'https://raw.githubusercontent.com/insaid2018/Domain_Case_Studies/master/Healthcare/Heart%20Disease%20Diagnosis/Data/heart_disease.csv'

replaceMissingMean(link=url, datacolnames=None, colname='cholesterol')

Before Replacing, Contains Missing?: True
After Replacing, Contains Missing?: False


age  sex   chest_pain_type  ...  num_major_vessels  thalassemia  target
0     63    1      asymptomatic  ...                  0            1       1
1     37    1  non-anginal pain  ...                  0            2       1
2     41    0   atypical angina  ...                  0            2       1
3     56    1   atypical angina  ...                  0            2       1
4     57    0   typical anginia  ...                  0            2       1
..   ...  ...               ...  ...                ...          ...     ...
361   45    1      asymptomatic  ...                  0            3       0
362   48    1    typical angina  ...                  0            3       0
363   66    1    typical angina  ...                  0            1       1
364   34    1      asymptomatic  ...                  0            2       1
365   51    1    typical angina  ...                  3            3       0

[366 rows x 14 columns]

---
**<h4>Question 10:** Create a function that generates random numbers ranging from Date A to Date B.</h4>

- For example, you passed `start_date = "2016-01-01"` and `end_date = "2016-12-31"`.

- The frequency used will be `freq = "B"` Here "B" refers to the business day frequency.

- The output data will be as shown below:

|Date|Value|
|:--:|:--:|
|2016-01-01|0.095|
|2016-01-02|0.340|
|2016-01-03|0.253|
|2016-01-04|0.254|
|2016-01-05|0.243|
||...|
|2016-12-30|0.664|
|2016-12-31|0.343|

<details>

**<summary>Hint:</summary>**

- Take help from pandas date_range functionality.

- Set all the parameters according to the function requirement.

- Generate data of length equal to the date length.

- Create a pandas series out of the random data and date and return it.

</details>

In [20]:
def randDateData(start_date=None, end_date=None, freq=None):
  """Generate random values ranging from Date A to Date B."""

  # Generates date ranging from specified value in the function
  date = pd.date_range(start=start_date, end=end_date, freq=freq)

  # Generate random data ranging from Date A to Date B
  data = np.random.rand(len(date))

  # Create series of random values from the generated date
  rand_date_val = pd.Series(data=data, index=date)

  return rand_date_val

In [21]:
randDateData(start_date='2016-01-01', end_date='2016-12-31', freq="B")

2016-01-01    0.139002
2016-01-04    0.278280
2016-01-05    0.539974
2016-01-06    0.315909
2016-01-07    0.335295
                ...   
2016-12-26    0.108642
2016-12-27    0.218235
2016-12-28    0.531132
2016-12-29    0.672738
2016-12-30    0.520354
Freq: B, Length: 261, dtype: float64

---
**<h4>Question 11:** Create a function that performs one hot encoding over the data column.</h4>

- For example, you passed the dictionary or read the data and dataframe looks like as shown below:

|#|Animal|Age|Visits|Priority|
|:--:|:--:|:--:|:--:|:--:|
|01|dog|NaN|3|yes|
|02|dog|5|2|no|
|03|cat|2|3|no|
|04|snake|4|1|no|
|05|cat|NaN|1|yes|

- You column selected for one hot encoding is Animal.

- The returned data will look like as follow:

|#|Age|Visits|Priority|Animal_cat|Animal_dog|Animal_snake|
|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
|01|NaN|3|yes|0|1|0|
|02|5|2|no|0|1|0|
|03|2|3|no|1|0|0|
|04|4|1|no|0|0|1|
|05|NaN|1|yes|1|0|0|

<details>

**<summary>Hint:</summary>**

- There are many ways to accomplish this question.

- Either you can use a library or use pandas functionality of get_dummies() method.

- You can also add a condition in the function to identify the type of data and take actions accordingly.

- To know more about the one hot encoding click <a href="https://machinelearningmastery.com/why-one-hot-encode-data-in-machine-learning/">**here**</a>.

</details>

In [22]:
def oneHotEncode(data=None, cols=None):
  """Performs one hot encoding over the specified data column."""

  # Check if the data type is pandas dataframe
  if type(data) == 'pandas.core.frame.DataFrame':

    # Return the one hot encoded data column
    return pd.get_dummies(data=data, columns=cols)

  # Creates dataframe out of dictionary if data type is not dataframe
  data = pd.DataFrame(data=data)

  return pd.get_dummies(data=data, columns=cols)

In [23]:
data_dict = {'Animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
             'Age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3], 
             'Visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
             'Priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

oneHotEncode(data=data_dict, cols=['Animal'])

Age  Visits Priority  Animal_cat  Animal_dog  Animal_snake
0  2.5       1      yes           1           0             0
1  3.0       3      yes           1           0             0
2  0.5       2       no           0           0             1
3  NaN       3      yes           0           1             0
4  5.0       2       no           0           1             0
5  2.0       3       no           1           0             0
6  4.5       1       no           0           0             1
7  NaN       1      yes           1           0             0
8  7.0       2       no           0           1             0
9  3.0       1       no           0           1             0

---
**<h4>Question 12:** Create a function that removes duplicates (if present) and subtract the filtered data with the mean of the column.</h4>

- For example you have the following data:

|Column data|
|:--:|
|1|
|2|
|2|
|3|
|4|
|5|
|5|
|5|
|6|
|7|

- The filtered data will be:

|Column data|
|:--:|
|1|
|2|
|3|
|4|
|5|
|6|
|7|

- Let's say the mean of the column is "4", then the final series will be:

|Column data|
|:--:|
|-3|
|-2|
|-1|
|0|
|1|
|2|
|3|


<details>

**<summary>Hint:</summary>**

- Check if the passed data is a dataframe or not.

- If dataframe, then drop the duplicates using the drop_duplicates() functionality of pandas.

- Then subtract the output with the mean of the column and return it.

- If the type of data is dictionary, then typecast it to the pandas dataframe first.

- Drop the duplicates using the drop_duplicates() functionality of pandas.

- Then subtract the output with the mean of the column and return it.

</details>

In [24]:
def filterData(data=None, names=None):
  """Remove duplicates (if present) and subtract the filtered data with the mean of the column."""
  
  # Check if the data type is pandas dataframe
  if type(data) == 'pandas.core.frame.DataFrame':

    # Drop duplicates
    data.drop_duplicates(inplace=True)

    # Return the subtracted values
    return data.sub(other=data.mean(axis=0), axis=1)

  # Creates dataframe out of dictionary if data type is not dataframe
  data = pd.DataFrame(data=data, columns=names)

  # Drop duplicates
  data.drop_duplicates(inplace=True)

  # Return the subtracted values
  return data.sub(other=data.mean(axis=0), axis=1)

In [25]:
data = {'Column data': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]}
filterData(data=data, names=None)

Column data
0         -3.0
1         -2.0
3         -1.0
4          0.0
5          1.0
8          2.0
9          3.0

---
**<h4>Question 13:** Create a function that performs dataframe merging over the specified column name.</h4>

- For example,

<center><img src="https://pandas.pydata.org/pandas-docs/stable/_images/merging_merge_on_key_multiple.png"></center>

<details>

**<summary>Hint:</summary>**

- Check if the passed data are of type DataFrame.

- If yes, then merge data 1 and data 2 over the specified column name.

- To achive the merging, you can use pandas's merge() functionality.

- Next, if the data is of type dictionary, then typecast the data to DataFrame.

- Then merge data 1 and data 2 over the specified column name using merge() method.

</details>

In [26]:
def dataframeMerge(data1=None, data2=None, key=None):
  """Perform dataframe merging over the specified column name."""

  # Check if the data is of type DataFrame
  if (type(data1) == 'pandas.core.frame.DataFrame') & (type(data2) == 'pandas.core.frame.DataFrame'):

    # Return the merged data over the specified column name
    return pd.merge(left=data1, right=data2, on=key)

  # Typecast dictionary to dataframe
  data1 = pd.DataFrame(data=data1)
  data2 = pd.DataFrame(data=data2)

  # Return the merged data over the specified column name
  return pd.merge(left=data1, right=data2, on=key)

In [27]:
d1 = pd.DataFrame({"key1": ["K0", "K0", "K1", "K2"], "key2": ["K0", "K1", "K0", "K1"],
                     "A": ["A0", "A1", "A2", "A3"], "B": ["B0", "B1", "B2", "B3"],
                     })

d2 = pd.DataFrame({"key1": ["K0", "K1", "K1", "K2"], "key2": ["K0", "K0", "K0", "K0"],
                   "C": ["C0", "C1", "C2", "C3"], "D": ["D0", "D1", "D2", "D3"],
                   })

dataframeMerge(data1=d1, data2=d2, key=["key1", "key2"])

key1 key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K1   K0  A2  B2  C1  D1
2   K1   K0  A2  B2  C2  D2

---
**<h4>Question 14:** Create a function that returns all rows of a DataFrame containing a given substring.</h4>

- For example, you passed 9 random values such as:
  -  `[0.83086063, 0.00536904, 0.58981977, 0.76704028, 0.55501003, 0.91354796, 0.57879962, 0.76508311, 0.66402454]`.

- The value returned will be:
    - `|0.83086063 0.00536904 0.58981977|`
    - `|0.76704028 0.55501003 0.91354796|`
    - `|0.57879962 0.76508311 0.66402454|`

<details>

**<summary>Hint:</summary>**

- Check if the type of data passed is of type DataFrame or not.

- If yes, then return the dataframe with filtered values containing substring in a specific column.

- You can use .str() along with the contains() method to filter out the data.

- If data is of type dictionary, then typecast it to DataFrame first.

- Then filter dataframe rows containing substring in a specific column.

</details>

In [28]:
def filterOnSubstring(data=None, colname=None, substring=None):
  """Filters dataframe containing a given substring."""

  # Check if the data is of type DataFrame
  if type(data) == 'pandas.core.frame.DataFrame':

    # Returns dataframe containing a specified substring
    return data[data[colname].str.contains(substring)]

  # Typecast to dataframe
  data = pd.DataFrame(data=data)
  
  # Returns dataframe containing a specified substring
  return data[data[colname].str.contains(substring)]

In [29]:
data = pd.DataFrame({'Name': ['Mukesh', 'Mohit', 'Hemanka', 'Abhigyan', 'Ashish'], 
                   'Team': ['Research', 'Research', 'Research', 'Research', 'Research'], 
                   'Position': ['Senior', 'Senior', 'Junior', 'Junior', 'Junior'], 
                   'Age': [24, 25, 23, 24, 24], 
                   'College': ['DAV University', np.nan, np.nan, np.nan, 'Lovely University'], 
                   'Salary': [30000, 30000, 50000, 10000, 10000]})

filterOnSubstring(data=data, colname='Position', substring='Junior')

Name      Team Position  Age            College  Salary
2   Hemanka  Research   Junior   23                NaN   50000
3  Abhigyan  Research   Junior   24                NaN   10000
4    Ashish  Research   Junior   24  Lovely University   10000

---
**<h4>Question 15:** Create a function that replaces values in a pandas dataframe using a regex.</h4>

- **Note:** If you don't know what is regular expression or regex, then click <a href="https://www.w3schools.com/python/python_regex.asp#:~:text=A%20RegEx%2C%20or%20Regular%20Expression,contains%20the%20specified%20search%20pattern.">**here**</a> to get more information.

- For example, you passed the dataframe that looks like as shown below:

|#|City|Event|Cost|
|:--:|:--:|:--:|:--:|
|01|New York|Music|10000|
|02|Parague|Poetry|5000|
|03|New Delhi|Theatre|15000|
|04|Venice|Comedy|2000|
|05|New Orleans|Tech Summit|12000|

- The output data will be as shown below:

|#|City|Event|Cost|
|:--:|:--:|:--:|:--:|
|01|New_York|Music|10000|
|02|Parague|Poetry|5000|
|03|New_Delhi|Theatre|15000|
|04|Venice|Comedy|2000|
|05|New_Orleans|Tech Summit|12000|

<details>

**<summary>Hint:</summary>**

- Check if the data passed is of type DataFrame or not.

- If yes, then replace the dataframe values with the specified parameters.

- The regex value will be specified in "to_replace" and the "value" to be replaced with.

- You can use .replace() functionality to accomplish this tasks.

- Next, if the type of data is not of type DataFrame, then typecast it to a DataFrame first.

- Then return the data with replaced value using .replace() method by specifying the regex.

</details>

---
**<h4>Question 16:** Create a function that stacks or unstack the data present in a dataframe.</h4>

- For example, you passed the dataframe that looks like as shown below:

|#|City|Event|Cost|
|:--:|:--:|:--:|:--:|
|0|New York|Music|10000|
|1|Parague|Poetry|5000|
|2|New Delhi|Theatre|15000|
|3|Venice|Comedy|2000|
|4|New Orleans|Tech Summit|12000|

- The condition passed is "stack" and the output data will be returned as shown below:

|#|Feature|Value|
|:--:|:--:|:--:|
|0|City|New York|
|0|Event|Music|
|0|Cost|10000|
|1|City|Parague|
|1|Event|Poetry|
|1|Cost|5000|
|2|City|New Delhi|
|2|Event|Theatre|
|2|Cost|15000|
|3|City|Venice|
|3|Event|Comedy|
|3|Cost|2000|
|4|City|New Orleans|
|4|Event|Tech Summit|
|4|Cost|12000|

<details>

**<summary>Hint:</summary>**

- Check if the data passed is of type DataFrame or not.

- If yes, then check if the passed condition is "unstack".

- If yes, then return the unstacked version of data using .unstack() method.

- Else return the stacked version of data using .stack() method.

- Next, if the data is not of type DataFrame, then typecast the passed data to a DataFrame.

- Check if the passed condition is "unstack".

- If yes, then return the unstacked version of data using .unstack() method.

- Else return the stacked version of data using .stack() method.

</details>

In [30]:
def stackOrUnstack(data=None, condition=None):
  """Returns a stacked or unstacked format of data in a dataframe."""

  # Check if the data is of type DataFrame
  if type(data) == 'pandas.core.frame.DataFrame':
    
    # Check if the condition is of type unstack
    if condition == 'unstack':
      return data.unstack()

    # Stack the data if the above condition fails to execute
    else:
      return data.stack()
  
  # Typecast the data to pandas dataframe
  data = pd.DataFrame(data=data)

  # Check if the condition is of type unstack
  if condition == 'unstack':
    return data.unstack()
    
  # Stack the data if the above condition fails to execute
  return data.stack() 

In [31]:
df = pd.DataFrame({'City':['New York', 'Parague', 'New Delhi', 'Venice', 'new Orleans'], 
                   'Event':['Music', 'Poetry', 'Theatre', 'Comedy', 'Tech Summit'], 
                   'Cost':[10000, 5000, 15000, 2000, 12000]})

stackOrUnstack(data=df, condition='stack')

0  City        New York
   Event          Music
   Cost           10000
1  City         Parague
   Event         Poetry
   Cost            5000
2  City       New Delhi
   Event        Theatre
   Cost           15000
3  City          Venice
   Event         Comedy
   Cost            2000
4  City     new Orleans
   Event    Tech Summit
   Cost           12000
dtype: object

---
**<h4>Question 17:** Create a function that returns a list of pandas dataframe.</h4>

- For example, you passed the dataframe that looks like as shown below:

|#|City|Event|Cost|
|:--:|:--:|:--:|:--:|
|0|New York|Music|10000|
|1|Parague|Poetry|5000|
|2|New Delhi|Theatre|15000|
|3|Venice|Comedy|2000|
|4|New Orleans|Tech Summit|12000|

- The returned output will be as shown below:

<center>[['New York', 'Music', 10000],</center>
<center>['Parague', 'Poetry', 5000],</center>
<center>['New Delhi', 'Theatre', 15000],</center>
<center>['Venice', 'Comedy', 2000],</center>
<center>['new Orleans', 'Tech Summit', 12000]]</center>

<details>

**<summary>Hint:</summary>**

- Check if the data passed is of type DataFrame or not.

- If yes, then iterate over the rows using .iterrows() functionality of pandas.

- Using conditional looping extract the index and rows of dataframe.

- Next, you need to use list comprehension to extract each value of a record to a list and return it back.

- If the passed data is of type other than DataFrame, then typecast it to a dataframe first.

- Initialize an empty list and, then iterate over the rows using .iterrows() functionality of pandas.

- Using conditional looping extract the index and rows of dataframe.

- Next, use a list comprehension to extract each value of a record to a list and append it to the initialized list.

- Return the appended list and display the output.

</details>

In [32]:
def listFromDataFrame(data=None):
  """Returns a list of rows from the pandas dataframe."""

  # Check if the data is of type DataFrame
  if type(data) == 'pandas.core.frame.DataFrame':
    
    # Retrieve index and rows of pandas dataframe
    for index, rows in data.iterrows():

      # Extract each value of cell of each row to a list
      row_list = [rows[i] for i in data.columns]

      # Return the row as a list
      return row_list
  
  # Typecast the data to a pandas dataframe
  data = pd.DataFrame(data=data)

  # Initiate an empty list
  row_list = []

  # Retrieve index and rows of pandas dataframe
  for index, rows in data.iterrows():

    # Extract each value of cell of each row to a list
    temp_list = [rows[i] for i in data.columns]

    # Append the temporary list of rows to the initialized list
    row_list.append(temp_list)

  # Return rows as a list
  return row_list

In [33]:
df = pd.DataFrame({'City':['New York', 'Parague', 'New Delhi', 'Venice', 'new Orleans'], 
                   'Event':['Music', 'Poetry', 'Theatre', 'Comedy', 'Tech Summit'], 
                   'Cost':[10000, 5000, 15000, 2000, 12000]})

listFromDataFrame(data=df)

[['New York', 'Music', 10000],
 ['Parague', 'Poetry', 5000],
 ['New Delhi', 'Theatre', 15000],
 ['Venice', 'Comedy', 2000],
 ['new Orleans', 'Tech Summit', 12000]]